src/core/shape_inference/include/gather_nd_shape_inference.hpp

```diff
@@ -0,0 +1,93 @@
+// Copyright (C) 2018-2023 Intel Corporation
+// SPDX-License-Identifier: Apache-2.0
+//
+#pragma once
+#include "openvino/op/gather_nd.hpp"
+#include "utils.hpp"
+
+namespace ov {
+namespace op {
+
+namespace gather_nd {
+template <class TShape, class TOp>
+std::vector<TShape> gather_nd_base_shape_infer(const TOp* op, const std::vector<TShape>& input_shapes) {
+    NODE_VALIDATION_CHECK(op, input_shapes.size() == 2);
+
+    const auto& data_pshape = input_shapes[0];
+    const auto& indices_pshape = input_shapes[1];
+
+    if (data_pshape.rank().is_static()) {
+        NODE_VALIDATION_CHECK(op, data_pshape.size() > 0, "Data rank must be at least 1.");
+        NODE_VALIDATION_CHECK(op,
+                              data_pshape.size() > op->get_batch_dims(),
+                              "Number of batch dimensions must not exceed a rank of data.");
+    }
+
+    if (indices_pshape.rank().is_static()) {
+        NODE_VALIDATION_CHECK(op, indices_pshape.size() > 0, "Indices rank must be at least 1.");
+        NODE_VALIDATION_CHECK(op,
+                              indices_pshape.size() > op->get_batch_dims(),
+                              "Number of batch dimensions must not exceed a rank of indices.");
+    }
+
+    if (data_pshape.rank().is_static() && indices_pshape.rank().is_static() &&
+        indices_pshape[indices_pshape.size() - 1].is_static()) {
+        auto batch_dims = op->get_batch_dims();
+        auto indices_tuple_length = indices_pshape[indices_pshape.size() - 1].get_length();  // last dim of indices
+
+        NODE_VALIDATION_CHECK(
+            op,
+            static_cast<int64_t>(indices_tuple_length + op->get_batch_dims()) <= data_pshape.rank().get_length(),
```

Comment:
Is cast required ? Is not possible to make both sides same type?
If not use `cmp::le` to safe compare different types e.g:
```cpp
cmp::le(indices_tuple_length + op->get_batch_dims(), data_pshape.rank().get_length())
```

Reply:
Updated to use `cmp::le`, applied in PR: https://github.com/openvinotoolkit/openvino/pull/15416